File: database/src/main/java/com/github/st1hy/countthemcalories/core/rx/Schedulers.java
package com.github.st1hy.countthemcalories.core.rx;

import rx.Scheduler;

public class Schedulers {
    private static final Hook DEFAULT = new HookImp();
    private static Hook hook = DEFAULT;

    public static Scheduler io() {
        return hook.io();
    }

    public static Scheduler computation() {
        return hook.computation();
    }

    public static Scheduler immediate() {
        return hook.immediate();
    }


    public static void registerHook(Hook hook) {
        Schedulers.hook = hook;
    }

    public static void reset() {
        Schedulers.hook = DEFAULT;
    }

    public interface Hook {
        Scheduler io();
        Scheduler computation();
        Scheduler immediate();
    }

    public static class HookImp implements Hook {
        @Override
        public Scheduler io() {
            return rx.schedulers.Schedulers.io();
        }

        @Override
        public Scheduler computation() {
            return rx.schedulers.Schedulers.computation();
        }

        @Override
        public Scheduler immediate() {
            return rx.schedulers.Schedulers.immediate();
        }
    }
}

File: app/src/main/java/com/github/st1hy/countthemcalories/utils/Schedulers.kt
package com.github.st1hy.countthemcalories.utils

import com.github.st1hy.countthemcalories.contract.Schedulers
import com.github.st1hy.countthemcalories.contract.SchedulersProvider
import rx.Scheduler
import rx.android.schedulers.AndroidSchedulers
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class SchedulersProviderImp @Inject constructor() : SchedulersProvider {

    private var hook: Schedulers = DEFAULT

    override fun io(): Scheduler {
        return hook.io()
    }

    override fun computation(): Scheduler {
        return hook.computation()
    }

    override fun ui(): Scheduler {
        return hook.ui()
    }

    override fun registerOverride(schedulers: Schedulers) {
        hook = schedulers
    }

    override fun reset() {
        hook = DEFAULT
    }
}

private val DEFAULT = HookImp()

class HookImp : Schedulers {
    override fun io(): Scheduler {
        return rx.schedulers.Schedulers.io()
    }

    override fun computation(): Scheduler {
        return rx.schedulers.Schedulers.computation()
    }

    override fun ui(): Scheduler {
        return AndroidSchedulers.mainThread()
    }
}